sqlglot/optimizer/projection_pushdown.py

```diff
@@ -26,137 +24,54 @@ def projection_pushdown(expression):
         sqlglot.Expression: optimized expression
     """
     expression = expression.copy()
-    _pushdown_statement(expression, SELECT_ALL)
-    return expression
-
-
-def _pushdown_statement(expression, parent_selections):
-    """
-    Search SELECT or UNION for columns that can be removed.
 
-    Args:
-        expression (exp.Select or exp.Union): expression to search
-        parent_selections: columns being selected by an outer query.
-            This can be the special value `SELECT_ALL`, which mean the outer query
-            is selecting everything.
-    Returns:
-        dict: Mapping of selectable names to columns.
-        This is used during recursion, so the outer query can:
-            1. pullup any selected columns
-            2. pushdown selected columns into CTEs
-    """
-    if isinstance(expression, exp.Select):
-        return _pushdown_select(expression, parent_selections)
-    if isinstance(expression, exp.Union):
-        if expression.args.get("distinct"):
-            # We can't remove selections on simple UNION
-            parent_selections = SELECT_ALL
+    # Map of Scope to all columns being selected by outer queries.
+    referenced_columns = defaultdict(set)
 
-        selections = _pushdown_select(expression.this, parent_selections)
-        selections = _merge_selections(
-            selections,
-            _pushdown_select(expression.args.get("expression"), parent_selections),
-        )
+    # We build the scope tree (which is traversed in DFS postorder), then iterate
+    # over the result in reverse order. This should ensure that the set of selected
+    # columns for a particular scope are completely build by the time we get to it.
+    for scope in reversed(list(traverse_scope(expression))):
+        parent_selections = referenced_columns.get(id(scope), {SELECT_ALL})
```

Comment:
why do you need to get id(scope), shouldn't it default to that? scope is not an expression so it doesn't overide hash right

Reply:
Good catch. Updating...